### PR TITLE
Proposals integrated in the main smart contract - first draft

### PR DESCRIPTION
**Pull request by karonka**

### CODE REVIEW

File: contracts/koh/contracts/KingAutomaton.sol [hunk 1]
```diff
@@ -178,6 +185,29 @@ contract KingAutomaton {
     }
 
     payForGas(id, 1);
+    return id;
+  }
+
+  function createProposal(address payable contributor, string memory title, string memory documents_link,
+      bytes memory documents_hash, uint256 contest_period, int256 approval_percentage, int256 contest_percentage)
```
Comment (akovachev) [inline contracts/koh/contracts/KingAutomaton.sol]: ``uint256 contest_period, int256 approval_percentage, int256 contest_percentage``
let's make these global parameters for now

File: contracts/koh/contracts/KingAutomaton.sol [hunk 2]
```diff
@@ -178,6 +185,29 @@ contract KingAutomaton {
...
+      public returns (uint256) {
```
Comment (akovachev) [inline contracts/koh/contracts/KingAutomaton.sol]: make it ``external`` and ``memory`` -> ``calldata`` for the parameters

File: contracts/koh/contracts/KingAutomaton.sol [hunk 3]
```diff
@@ -191,34 +221,30 @@ contract KingAutomaton {
 
   // Pay for multiple slots at once, 32 seems to be a reasonable amount.
   function payForGas(uint256 ballot_box_id, uint256 _slotsToPay) public validBallotBoxID(ballot_box_id) {
-    BallotBox storage p = ballot_boxes[ballot_box_id];
-    require((slots.length - p.paidSlots) >= _slotsToPay, "Too many slots!");
-    uint _newLength = p.paidSlots + _slotsToPay;
+    BallotBox storage b = ballot_boxes[ballot_box_id];
+    require((slots.length - b.paidSlots) >= _slotsToPay, "Too many slots!");
+    uint _newLength = b.paidSlots + _slotsToPay;
```
Comment (akovachev) [inline contracts/koh/contracts/KingAutomaton.sol]: Also pre-fetch these two in memory to save on gas costs
```
  uint256 _votesWords = b.votesWords;
  uint256 _slotsLength = slots.length;
```

File: contracts/koh/contracts/KingAutomaton.sol [hunk 4]
```diff
@@ -258,10 +286,55 @@ contract KingAutomaton {
...
 
+  function completeProposal(uint256 id) private activeBallot(id) {
+    Proposal storage p = proposals[id];
+    require(p.state == ProposalState.Started || p.state == ProposalState.Contested, "Invalid proposal state!");
+    BallotBox storage b = ballot_boxes[id];
+    p.state = ProposalState.Completed;
+    b.state = BallotBoxState.Inactive;
+  }
+
+  function updateProposalState(uint256 id) public validBallotBoxID(id) {
```
Comment (akovachev) [inline contracts/koh/contracts/KingAutomaton.sol]: This function only needs to check 3 proposal states:
- Started
- Accepted
- Contested

Let's split the logic in 3 sections (simulate a switch with if/else logic) and get rid of the requires.

File: contracts/koh/contracts/KingAutomaton.sol [hunk 5]
```diff
@@ -191,34 +221,30 @@ contract KingAutomaton {
...
+    if (b.paidSlots == slots.length) {
+        b.state = BallotBoxState.Active;
     }
   }
 
-  function calcVoteDifference(uint256 ballot_box_id) view private validBallotBoxID(ballot_box_id) returns (uint256) {
+  function calcVoteDifference(uint256 ballot_box_id) view private validBallotBoxID(ballot_box_id) returns (int256) {
     BallotBox storage p = ballot_boxes[ballot_box_id];
     uint256 yes = p.voteCount[1];
     uint256 no = p.voteCount[2];
```
Comment (akovachev) [inline contracts/koh/contracts/KingAutomaton.sol]:
```
    int256 yes = int256(p.voteCount[1]);
    int256 no = int256(p.voteCount[2]);
```

File: contracts/koh/contracts/KingAutomaton.sol (hunk 3 above)
Comment (karonka) [reply contracts/koh/contracts/KingAutomaton.sol]: First one is unnecessary

File: contracts/koh/contracts/KingAutomaton.sol [hunk 6]
```diff
@@ -190,76 +211,135 @@ contract KingAutomaton {
...
+    ProposalState p_state = p.state;
+    require(p_state == ProposalState.Started || p_state == ProposalState.Contested, "Invalid proposal state!");
+    p.state = ProposalState.Completed;
+    b.state = BallotBoxState.Inactive;
+  }
+
+  function updateProposalState(uint256 _id) public validBallotBoxID(_id) {
+    Proposal storage p = proposals[_id];
+    ProposalState p_state = p.state;
+    if (p_state == ProposalState.Uninitialized) {  // There is no proposal linked to this ballot
```
Comment (akovachev) [inline contracts/koh/contracts/KingAutomaton.sol]: Get rid of this check - if proposal is Uninitialized, then nothing is going to be done anyways.

File: contracts/koh/test/ProposalsTest.js [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+    await koh.updateProposalState(id);
+    // States shouldn't change during initial time
+    ballot = await koh.ballot_boxes(id);
+    ballot_state = ballot.state;
+    assert.equal(ballot_state, 2, "Ballot box state is not Active!");
+    proposal = await koh.proposals(id)
+    proposal_state = proposal.state;
+    assert.equal(proposal_state, 1, "Proposal state is not Started!");
+    await await web3.currentProvider.send(
+        {jsonrpc: "2.0", method: "evm_increaseTime", params: [proposal_start_period], id: 0}, (err, result) => {
```
Comment (akovachev) [inline contracts/koh/test/ProposalsTest.js]: Can you add this to the utils.js as a function, so it's just one line of code here in the tests?

File: contracts/koh/test/ProposalsTest.js [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+    assert.equal(ballot_state, 3, "Ballot box state is not Inactive!");
+    let proposal = await koh.proposals(id)
+    let proposal_state = proposal.state;
+    assert.equal(proposal_state, 3, "Proposal state is not Rejected!");
+    let vote_difference = await koh.calcVoteDifference(id);
+    assert.equal(vote_difference, 0, "Vote difference is incorrect!");
+  });
+
+  // 2 positive 1 negative // 25% approval
+  it("approval then refection during initial voting", async() => {
```
Comment (akovachev) [inline contracts/koh/test/ProposalsTest.js]: typo: refection -> rejection

File: contracts/koh/test/ProposalsTest.js [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+    await koh.payForGas(id, 3);
+
+    // Cast 2 positive votes
+    await koh.castVote(id, 0, 1);
+    await koh.castVote(id, 1, 1);
+    // Cast negative vote
+    await koh.castVote(id, 2, 2);
+
+    await await web3.currentProvider.send(
+        {jsonrpc: "2.0", method: "evm_increaseTime", params: [proposal_start_period], id: 0}, (err, result) => {
```
Comment (akovachev) [inline contracts/koh/test/ProposalsTest.js]: Same here -> extract to a function in utils.js

File: contracts/koh/test/ProposalsTest.js [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+    ballot = await koh.ballot_boxes(id);
+    ballot_state = ballot.state;
+    assert.equal(ballot_state, 2, "Ballot box state is not Active!");
+    proposal = await koh.proposals(id)
+    proposal_state = proposal.state;
+    assert.equal(proposal_state, 4, "Proposal state is not Contested!");
+
+    // Wait for contest to end
+    await await web3.currentProvider.send(
+        {jsonrpc: "2.0", method: "evm_increaseTime", params: [contest_period], id: 0}, (err, result) => {
```
Comment (akovachev) [inline contracts/koh/test/ProposalsTest.js]: and here

File: contracts/koh/test/ProposalsTest.js [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+    ballot_state = ballot.state;
+    assert.equal(ballot_state, 2, "Ballot box state is not Active!");
+    proposal = await koh.proposals(id)
+    proposal_state = proposal.state;
+    assert.equal(proposal_state, 4, "Proposal state is not Contested!");
+
+
+    // Wait for contest to end
+    await await web3.currentProvider.send(
+        {jsonrpc: "2.0", method: "evm_increaseTime", params: [contest_period], id: 0}, (err, result) => {
```
Comment (akovachev) [inline contracts/koh/test/ProposalsTest.js]: here too

File: contracts/koh/test/ProposalsTest.js [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+    await koh.updateProposalState(id);
+    // States shouldn't change until the end of contest
+    ballot = await koh.ballot_boxes(id);
+    ballot_state = ballot.state;
+    assert.equal(ballot_state, 2, "Ballot box state is not Active!");
+    proposal = await koh.proposals(id)
+    proposal_state = proposal.state;
+    assert.equal(proposal_state, 4, "Proposal state is not Contested!");
+
+
```
Comment (akovachev) [inline contracts/koh/test/ProposalsTest.js]: 2 empty lines -> 1 empty line

File: contracts/koh/contracts/KingAutomaton.sol [hunk 13]
```diff
@@ -105,6 +105,12 @@ contract KingAutomaton {
   uint256 constant BITS_PER_VOTE = 2;
   uint256 constant VOTES_PER_WORD = 127;
   uint256 constant BASE_MASK = (1 << BITS_PER_VOTE) - 1;
+  uint256 constant PROPOSAL_START_PERIOD = 90 seconds; // 1 weeks;
+  uint256 constant CONTEST_PERIOD = 90 seconds;  //
+  int256 constant APPROVAL_PERCENTAGE = 50;
```
Comment (akovachev) [inline contracts/koh/contracts/KingAutomaton.sol]: Make APPROVAL_PERCENTAGE / CONTEST_PERCENTAGE parameters of the constructor, so we can test with different values.

File: contracts/koh/test/ProposalsTest.js [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+  it("contested then approval", async() => {
+    await koh.setOwnerAllSlots();
+    await koh.payForGas(id, 3);
+
+    // Cast 2 positive votes
+    await koh.castVote(id, 0, 1);
+    await koh.castVote(id, 1, 1);
+    await koh.updateProposalState(id);
+    await await web3.currentProvider.send(
```
Comment (akovachev) [inline contracts/koh/test/ProposalsTest.js]: Here too...

File: contracts/koh/test/ProposalsTest.js [hunk 15]
```diff
@@ -0,0 +1,271 @@
...
+  });
+
+  it("contested then rejection", async() => {
+    await koh.setOwnerAllSlots();
+    await koh.payForGas(id, 3);
+
+    // Cast 2 positive votes
+    await koh.castVote(id, 0, 1);
+    await koh.castVote(id, 1, 1);
+    await await web3.currentProvider.send(
```
Comment (akovachev) [inline contracts/koh/test/ProposalsTest.js]: Here too

File: contracts/koh/test/ProposalsTest.js [hunk 16]
```diff
@@ -0,0 +1,271 @@
...
+    await koh.payForGas(id, 3);
+    // Test initial vote values
+    let vote = await koh.getVote(id, 0);
+    let pos_vote_count = await koh.getVoteCount(id, 1);
+    let neg_vote_count = await koh.getVoteCount(id, 2);
+    assert.equal(vote, 0, "Incorrect initial vote!");
+    assert.equal(pos_vote_count, 0, "Incorrect initial positive vote count!");
+    assert.equal(neg_vote_count, 0, "Incorrect initial negative vote count!");
+    // Cast positive vote
+    await koh.castVote(id, 0, 1);
```
Comment (akovachev) [inline contracts/koh/test/ProposalsTest.js]: Make VOTE_YES and VOTE_NO constants you can use for readability throughout the test.
`await koh.castVote(id, 0, 1);` would become `await koh.castVote(id, 0, VOTE_YES);`

File: contracts/koh/test/ProposalsTest.js [hunk 17]
```diff
@@ -0,0 +1,271 @@
...
+    vote = await koh.getVote(id, 0);
+    pos_vote_count = await koh.getVoteCount(id, 1);
+    assert.equal(vote, 1, "Incorrect vote!");
+    assert.equal(pos_vote_count, 1, "Incorrect positive vote count!");
+    // Change vote
+    await koh.castVote(id, 0, 2);
```
Comment (akovachev) [inline contracts/koh/test/ProposalsTest.js]: `await koh.castVote(id, 0, 2);` -> `await koh.castVote(id, 0, VOTE_NO);`

**Pull request merged**